python/pylibraft/pylibraft/neighbors/cagra/cagra.pyx

```diff
@@ -0,0 +1,852 @@
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+import warnings
+
+import numpy as np
+
+from cython.operator cimport dereference as deref
+from libc.stdint cimport (
+    int8_t,
+    int32_t,
+    int64_t,
+    uint8_t,
+    uint32_t,
+    uint64_t,
+    uintptr_t,
+)
+from libcpp cimport bool, nullptr
+from libcpp.string cimport string
+
+from pylibraft.distance.distance_type cimport DistanceType
+
+from pylibraft.common import (
+    DeviceResources,
+    ai_wrapper,
+    auto_convert_output,
+    cai_wrapper,
+    device_ndarray,
+)
+from pylibraft.common.cai_wrapper import wrap_array
+from pylibraft.common.interruptible import cuda_interruptible
+
+from pylibraft.common.handle cimport device_resources
+
+from pylibraft.common.handle import auto_sync_handle
+from pylibraft.common.input_validation import is_c_contiguous
+
+from rmm._lib.memory_resource cimport (
+    DeviceMemoryResource,
+    device_memory_resource,
+)
+
+cimport pylibraft.neighbors.cagra.cpp.c_cagra as c_cagra
+cimport pylibraft.neighbors.ivf_flat.cpp.c_ivf_flat as c_ivf_flat
+cimport pylibraft.neighbors.ivf_pq.cpp.c_ivf_pq as c_ivf_pq
+from pylibraft.common.optional cimport make_optional, optional
+
+from pylibraft.neighbors.common import _check_input_array, _get_metric
+
+from pylibraft.common.cpp.mdspan cimport (
+    device_matrix_view,
+    device_vector_view,
+    make_device_vector_view,
+    row_major,
+)
+from pylibraft.common.mdspan cimport (
+    get_dmv_float,
+    get_dmv_int8,
+    get_dmv_int64,
+    get_dmv_uint8,
+    get_dmv_uint32,
+    get_hmv_float,
+    get_hmv_int8,
+    get_hmv_int64,
+    get_hmv_uint8,
+    get_hmv_uint32,
+    make_optional_view_int64,
+)
+from pylibraft.neighbors.common cimport _get_metric_string
+from pylibraft.neighbors.ivf_pq cimport ivf_pq
+from pylibraft.neighbors.ivf_pq.cpp.c_ivf_pq cimport (
+    index_params as IVFPQ_IP,
+    search_params as IVFPQ_SP,
+)
+
+
+cdef class IndexParams:
+    cdef c_cagra.index_params params
+
+    def __init__(self, *,
+                 metric="sqeuclidean",
+                 intermediate_graph_degree=128,
+                 graph_degree=64,
+                 add_data_on_build=True):
+        """"
+        Parameters to build index for IVF-PQ nearest neighbor search
+
+        Parameters
+        ----------
+        metric : string denoting the metric type, default="sqeuclidean"
+            Valid values for metric: ["sqeuclidean", "inner_product",
+            "euclidean"], where
+            - sqeuclidean is the euclidean distance without the square root
+              operation, i.e.: distance(a,b) = \\sum_i (a_i - b_i)^2,
+            - euclidean is the euclidean distance
+            - inner product distance is defined as
+              distance(a, b) = \\sum_i a_i * b_i.
+        intermediate_graph_degree : int, default = 128
+
+        graph_degree : int, default = 64
+
+        add_data_on_build : bool, default = True
```

Comment:
I don't think this parameter is part of cagra is it? This was something that was added to the IVF methods so the user could bypass extend if they wanted.